### PR TITLE
Integrate CSV parser to parse sample sets

### PR DESCRIPTION
**Pull request by ChristianKaltenecker**

### CODE REVIEW

File: include/vara/Sampling/SampleSetParser.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+#include <vector>
+
+namespace vara::sampling {
+
+class SampleSetParser {
+public:
+  /// This method parses the given csv file as a string. This csv file contains
+  /// configurations from a sample set.
+  /// \param Model the corresponding feature model
+  /// \param csv the path to the csv file as a string
```
Comment (vulder) [inline include/vara/Sampling/SampleSetParser.h:20]:
```suggestion
  ///
  /// \param Model the corresponding feature model
  /// \param csv the path to the csv file as a string
  ///
```

File: lib/Sampling/SampleSetParser.cpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      std::vector<std::unique_ptr<vara::feature::Configuration>>>();
+  csv::CSVReader Reader(CsvPath, Format);
+  for (const csv::CSVRow &Row : Reader) {
+    auto Config = std::make_unique<vara::feature::Configuration>();
+    auto ColumnNames = Row.get_col_names();
+    for (auto *F : Model.features()) {
+      if (std::find(ColumnNames.begin(), ColumnNames.end(), F->getName()) !=
+          ColumnNames.end()) {
+        if (F->getKind() == feature::Feature::FeatureKind::FK_BINARY ||
+            F->getKind() == feature::Feature::FeatureKind::FK_ROOT) {
```
Comment (vulder) [inline lib/Sampling/SampleSetParser.cpp]:
```suggestion
        if (auto Kind = F->getKind(); Kind == feature::Feature::FeatureKind::FK_BINARY ||
            Kind == feature::Feature::FeatureKind::FK_ROOT) {
```

File: unittests/Sampling/SampleSetParserTests.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+TEST(SampleSetParser, testCSVParsing) {
+  auto CsvPath = getTestResource("configs_dune.csv");
+
+  auto FMFS =
+      llvm::MemoryBuffer::getFileAsStream(getTestResource("test_dune_num.xml"));
+  EXPECT_TRUE(FMFS);
+  auto P = vara::feature::FeatureModelXmlParser(FMFS.get()->getBuffer().str());
+  auto FM = P.buildFeatureModel();
+  auto Configs = SampleSetParser::readConfigurations(*FM, CsvPath);
+  EXPECT_EQ(Configs->size(), 40);
```
Comment (vulder) [inline unittests/Sampling/SampleSetParserTests.cpp]:
```suggestion
  ASSERT_EQ(Configs->size(), 40);
```

File: unittests/Sampling/SampleSetParserTests.cpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  testConfigurationOption(FirstConfig, "SeqGS", "true");
+  testConfigurationOption(FirstConfig, "SeqSOR", "false");
+  testConfigurationOption(FirstConfig, "CGSolver", "false");
+  testConfigurationOption(FirstConfig, "BiCGSTABSolver", "false");
+  testConfigurationOption(FirstConfig, "LoopSolver", "false");
+  testConfigurationOption(FirstConfig, "GradientSolver", "true");
+  testConfigurationOption(FirstConfig, "pre", "0");
+  testConfigurationOption(FirstConfig, "post", "3");
+  testConfigurationOption(FirstConfig, "cells", "55");
+  // Check the configuration options of the last configuration
```
Comment (vulder) [inline unittests/Sampling/SampleSetParserTests.cpp]:
```suggestion
  testConfigurationOption(FirstConfig, "cells", "55");
  
  // Check the configuration options of the last configuration
```

File: unittests/Sampling/SampleSetParserTests.cpp [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  auto CsvPath = getTestResource("configs_dune.csv");
+
+  auto FMFS =
+      llvm::MemoryBuffer::getFileAsStream(getTestResource("test_dune_num.xml"));
+  EXPECT_TRUE(FMFS);
+  auto P = vara::feature::FeatureModelXmlParser(FMFS.get()->getBuffer().str());
+  auto FM = P.buildFeatureModel();
+  auto Configs = SampleSetParser::readConfigurations(*FM, CsvPath);
+  EXPECT_EQ(Configs->size(), 40);
+  // Check the configuration options of the first configuration
```
Comment (vulder) [inline unittests/Sampling/SampleSetParserTests.cpp:32]:
```suggestion

  // Check the configuration options of the first configuration
```

**Review by vulder — CHANGES_REQUESTED** [review]
Rest looks good.

Just one thing, I might have missed it. Where is the writing of the other config format for the TS? Or should we add it afterwards?

**Review by vulder — APPROVED** [review]

**Review by s9latimm — APPROVED** [review]

**Pull request merged**